### DISCUSSION TITLE
Multi-currency support

### DISCUSSION THREAD

**Discussion by nitorrep** [investbrainapp/investbrain#62] [Ideas]
Hi !
I've just tried you app and it seems really promising, do you plain to be able to handle other currency than USD?
I'm from Switzerland and not all my investments are in USD, so it would be great to be able to do so

**Comment by hackeresq** [comment]
Hi @nitorrep this is on the roadmap! (Discussion #15)

Still tinkering with how this will have to work from a data architecture perspective.  So far there have been two struggles with multi-currency support:
1. Some market data providers don't include the currency in their schemas, which could limit future extensibility
2. Summing up the totals between holdings traded in different currencies would require converting to a common currency first.

Would be interested to hear how the community would tackle these two considerations?

**Reply by MRobi1** [reply]
> 2. Summing up the totals between holdings traded in different currencies would require converting to a common currency first.

Chiming in here. Ghostfolio takes a common currency approach, in their case USD, but the way they've implemented it creates a fundamental accounting flaw and makes it completely unusable for anybody dealing in multiple currencies.

If I invest $1,000CAD in AAPL, my cost basis must **always** remain $1,000CAD. What Ghostfolio does is takes that $1,000CAD, converts to USD at the time of the transaction. It then takes that USD amount and converts it back to CAD based on today's exchange rate when calculating return. So my $1,000 CAD cost basis becomes $1,100 today, $950 tmr, $1,005 the next day, etc..

I think the only way to properly implement this is to store the buy transaction in the selected currency and convert the current market value to that selected currency and then run the calculations for return.

**Reply by hackeresq** [reply]
I've tinkered with various options here. Your option would work, but could lead to significant memory usage with larger portfolios. Particularly because the "math" happens outside the database during runtime.

The ghostfolio approach doesn't take into account backdated transactions and, as you note, performance always uses the current exchange rate - which is memory efficient at the cost of accuracy.

I landed on a combination of these solutions. Essentially I convert to a "base currency" but keep the original cost basis/sale price for display purposes. For calculations, I'll be recording daily exchange rates to the database and joining that data in the DB queries. This applies to transactions and dovidends. Letting the DB handle this "math" is slightly more memory hungry, but is much more optimized since it only hits the DB once. 

As part of the pre-release this week, I'll draw up a quick visual to show how this works. Might also do a quick blog post to demonstrate the pros and cons of each approach. This was a surprisingly tricky issue!

**Reply by MRobi1** [reply]
> For calculations, I'll be recording daily exchange rates to the database and joining that data in the DB queries. 

Overall I see this approach being fine because the ACB will stay constant which is the major flaw of ghostfolio. 

Just 1 question... How will that work for same-day transactions since daily exchange rates are only published daily at 16:30 ET? If I buy a security today at 13:00 and record it before the daily exchange rate is posted, what exchange rate will be used for that transaction? Will it be stored at current exchange rate for same-day transactions and daily exchange rate when entering historical transactions?

**Comment by nitorrep** [comment]
HI @hackerESQ ,
You should maybe "just" (I love saying that, it's never "just" that easy ) have exchanges rates stored (If you want to avoid to store all exchange rate you could rather have all echange rates against USD and then convert twice. For exemple you have a transaction in EUR in a CHF portfolio (From EUR -> USD then from USD -> CHF).

And after on your structure you might define a reference currency at portfolio level and convert transactions (Who could be in any currency) to this main currency, same for valuations.

**Comment by flowcool** [comment]
Any news about this one ?

**Reply by hackeresq** [reply]
This one is top priority for me right now!

**Comment by deanpribetic** [comment]
Hello - is this still planned? As someone not in the US I dunno if InvestBrain will work for me?

**Comment by hackeresq** [comment]
Yep, still planned!

Working through some final kinks but multi-currency will be in the next
release. I'll probably mark it as "pre-release" at first since it likely
will have breaking changes (though the plan is to limit those obviously).

**Comment by hackeresq** [comment]
Currently using Frankfurter <https://frankfurter.dev/> which is also open
source and scrapes data from the European Central Bank.  Intra-day exchange
rates will always be a compromise without using commercial "live" data.
Currently I have Investbrain syncing intra-day every 3 hours.

Transactions will be stored like this:

   1. Where `currency`of security != `base currency`
   2. Convert cost basis (and sale price) to `base currency` using current
   (or historic) rate based on date of transaction
   3. Store two versions of cost basis - first column is `*_base` (e.g.
   `base currency*_base*`) and the other in market currency (e.g.
   `cost_basis`) - the same thing is true for dividends where this historical
   problem is an issue

Then, when performing performance calculations:

   1. For each portfolio, join the `currency rates` table to transactions
   on `date` in a subquery
   2. Convert the transactions `base currency` to user's `display currency`
   in the subquery using the rate on the date of the transaction
   3. Sum up all transactions in the `display currency`

This performance query is likely going to get cached since it's slightly
"expensive".

But to your point, there's inevitably going to be edge cases where a rate
wildly swings intra-day. I expect these edge cases to be rare and have a
limited impact on overall performance calculations.

Hope that makes sense? And also open to trying any alternatives if I've
missed anything.






On Mon, Mar 31, 2025 at 11:00 AM Mike ***@***.***> wrote:

> For calculations, I'll be recording daily exchange rates to the database
> and joining that data in the DB queries.
>
> Overall I see this approach being fine because the ACB will stay constant
> which is the major flaw of ghostfolio.
>
> Just 1 question... How will that work for same-day transactions since
> daily exchange rates are only published daily at 16:30 ET? If I buy a
> security today at 13:00 and record it before the daily exchange rate is
> posted, what exchange rate will be used for that transaction? Will it be
> stored at current exchange rate for same-day transactions and daily
> exchange rate when entering historical transactions?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/investbrainapp/discussions/62#discussioncomment-12678626>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD63TF2CIP23YFVIDP2RFAD2XFRABAVCNFSM6AAAAABZV4OLUCVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTENRXHA3DENQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
Corey

**Comment by hackeresq** [comment]
As a quick update for the folks waiting for multi-currency - very close to publishing the pre-release version (hoping this weekend).

Here's what the UI will look like:

**Portfolio View:**

![image](https://github.com/user-attachments/assets/ea1c0d94-9d01-460c-b271-a6b19bb75e4a)

**Holding/Transaction View:**

![image](https://github.com/user-attachments/assets/dd68fe30-254c-4782-8ff7-a909b7ef355c)

**Settings page:**

<img width="1311" alt="image" src="https://github.com/user-attachments/assets/5ab805e9-7bfb-46bb-b78e-1162faca02be" />

**Comment by kanwarujjaval** [comment]
Will this allow me use to symbols that do not support USD?

example: [EXIDEIND.NS](https://finance.yahoo.com/quote/EXIDEIND.NS/)

**Reply by hackeresq** [reply]
That is certainly the goal! 👍 

The core is mostly ready, just writing some tests before tagging the pre-release.

We're close!

**Reply by kanwarujjaval** [reply]
That is amazing! Glad I found this project!
Really appreciate the great work here.

**Comment by hackeresq** [comment]
Great news y'all!  I've just tagged a pre-release which includes multi-currency support: [v1.2.0-alpha](https://github.com/investbrainapp/investbrain/releases/tag/v1.2.0-alpha).

Before spinning up the pre-release, I very much encourage creating a backup (or running separate from your production instance). I very much anticipate there will be bugs and issues. 

Interested in feedback!

**Reply by kanwarujjaval** [reply]
I was able to get it to work, but with mariadb

mysql Had this error

`SQLSTATE[42000]: Syntax error or access violation: 1101 BLOB, TEXT, GEOMETRY or JSON column 'options' can't have a default value`

Also 

The currency seems to be working, but I cannot import symbols for other currencies eg: [EXIDEIND.NS](https://finance.yahoo.com/quote/EXIDEIND.NS/)

is this by design?

**Reply by srikakulamts** [reply]
Having the same issue. Is there a patch available?

**Reply by hackeresq** [reply]
This is still pre-release and I'm making many tweaks. 

I've definitely identified some areas where memory usage spikes and causes issues syncing currency rates. More updates to come soon.

But based on your note, I can't really do much. Can you share more specifics on what issue you're facing and steps to reproduce?

**Reply by kanwarujjaval** [reply]
When I try to add a symbol from NSE (Indian Exchange, the symbol above EXIDEIND.NS) there is an error that symbol is not valid. It is a valid symbol from yahoo api.
It may be possible that scheb/yahoo-finance-api does not support these symbols?

**Reply by hackeresq** [reply]
THanks for the additional details. I'll work on reproducing this and do some troubleshooting this weekend.

**Comment by MRobi1** [comment]
Trying to test it out but can't get it running at all.
Using default compose file I'm getting the "Waiting for database" error. And when switching to sqlite I get the 500 server error

**Reply by kanwarujjaval** [reply]
Switching to mariadb in docker compose worked for me
https://github.com/orgs/investbrainapp/discussions/62#discussioncomment-12811498

**Reply by MRobi1** [reply]
> Switching to mariadb in docker compose worked for me https://github.com/orgs/investbrainapp/discussions/62#discussioncomment-12811498

Do you mind posting your working compose file?

**Reply by kanwarujjaval** [reply]
only made changes here

```yaml
mysql:
    image: mariadb:10.6
    container_name: investbrain-mysql
    restart: unless-stopped
    environment:
      MYSQL_DATABASE: ${DB_DATABASE:-investbrain}
      MYSQL_USER: ${DB_USERNAME:-investbrain}
      MYSQL_PASSWORD: ${DB_PASSWORD:-investbrain}
      MYSQL_ROOT_PASSWORD: ${DB_PASSWORD:-investbrain}
    volumes:
      - /home/user/investbrain/mysql:/var/lib/mysql
    networks:
      - investbrain-network
```

**Comment by MRobi1** [comment]
Testing out multi-currency and running into a few issues.
First, I was unable to go from :latest to :1.2.0-beta1. It gave me an error about today's profit. I tried loading alpha3 and alpha2 but it was giving me database write permissions. To get to 1.2.0-beta1 I had to load 1.2.0-alpha, create the user, then upgrade to 1.2.0=beta1

On both alpha and beta1 I'm unable to set Locale Options
![image](https://github.com/user-attachments/assets/0c3ba444-ed6b-4245-946e-be66610047f5)
Hitting save gives this error
![image](https://github.com/user-attachments/assets/78368e1d-a735-48e9-b678-6033df8cf102)

I'm also unable to add non NYSE listed equities
![image](https://github.com/user-attachments/assets/8438b152-450f-4172-92a3-a7185464ccb0)
Even if they're listed on Yahoo
https://ca.finance.yahoo.com/quote/XEQT.TO/?guccounter=1&guce_referrer=aHR0cHM6Ly93d3cuZ29vZ2xlLmNvbS8&guce_referrer_sig=AQAAAIUeVvv__-gRisP_5nS7ysHbmMDyV_VlWvr-yYUHxOD7ttsupvQ7DvCXqzh1mwHNkERQa1--1YpORECnix1I2sFGi41gFTwKHb-l2KOGGDYVh-XAk1Eu8DKTBmTsmsGTV_SbnqjeORYskQ_6UvaO-cPBz_TRS9vCsq19wPwsRGB7

Semi-unrelated question, in the add transactions screen, should there be an auto-complete where if I were to start typing AAPL for example, it will give a dropdown to select the right security? If so... this is also not working. If not.... feature request please! lol

**Comment by MRobi1** [comment]
Just testing 1.2.0-beta4 and still getting the same.

1 thing I've noticed... aapl will return the symbol provided is not valid, but AAPL is accepted. So it appears the symbol is case sensitive. But it also does not seem to accept Canadian ETF's such as XEQT (also XEQT.TO), ZEQT, XGRO, etc... I assume this has something to do with the data provider.

But I also have market provider set 
MARKET_DATA_PROVIDER: "yahoo,finnhub"
which should both support these symbols.

EDIT: Just tested and the ability to change locale is also still the same issue.

EDIT 2: I'm actually thinking this has something to do with the input form, and not the provider.
The error I receive is "The symbol provided (XEQT.TO) is not valid", but the symbol error in YahooMarketData.php is "Could not find ticker on Yahoo"

**Comment by MRobi1** [comment]
Just tried this on 1.2.1
I'm now able to change locale and set base currency to CAD.

But get an error for all tickers on the TSX. I've tried every combination of XEQT.TO I can think of and get a symbol is not valid.
https://ca.finance.yahoo.com/quote/XEQT.TO/

@hackeresq any thoughts?

**Reply by hackeresq** [reply]
Very likely this is related to the Yahoo 429 issue (#102).  

When you create a transaction, there is a validation check to ensure symbols are valid.  

Can you check your laravel.log file to see if the Yahoo driver is hitting the 429 error? 

There's a fix in the underlying package for the Yahoo driver, but it requires a bit of re-working the Docker image. I'm working on it now.

You can test out the multi currency capabilities on the hosted version: https://investbra.in/dashboard

**Reply by hackeresq** [reply]
Going to mark this discussion as closed since v1.2.0 adds multi-currency.  Let's move this conversation re: Yahoo 429 issue over to #102.
